File: frontend/src/lib/api/parent-dashboard.ts
/**
 * Parent Dashboard API client for dashboard operations.
 *
 * Uses Zod schemas for runtime validation of API responses to ensure
 * type safety and catch API contract changes early.
 */

import { api } from './client';
import {
  DashboardOverviewResponseSchema,
  StudentProgressResponseSchema,
  WeeklyInsightsResponseSchema,
  GoalListResponseSchema,
  GoalWithProgressResponseSchema,
  NotificationListResponseSchema,
  NotificationResponseSchema,
  NotificationPreferencesResponseSchema,
  type DashboardOverviewResponse as DashboardOverviewApiResponse,
  type StudentProgressResponse as StudentProgressApiResponse,
  type WeeklyInsightsResponse as WeeklyInsightsApiResponse,
  type GoalWithProgressResponse as GoalWithProgressApiResponse,
  type NotificationResponse as NotificationApiResponse,
  type NotificationPreferencesResponse as NotificationPreferencesApiResponse,
} from './schemas/parent-dashboard';

// =============================================================================
// Types
// =============================================================================

export interface DashboardStudentSummary {
  id: string;
  displayName: string;
  gradeLevel: number;
  schoolStage: string;
  frameworkId: string | null;
  totalXp: number;
  level: number;
  currentStreak: number;
  longestStreak: number;
  lastActiveAt: string | null;
  sessionsThisWeek: number;
  studyTimeThisWeekMinutes: number;
}

export interface DashboardOverview {
  students: DashboardStudentSummary[];
  totalStudyTimeWeekMinutes: number;
  totalSessionsWeek: number;
  unreadNotifications: number;
  activeGoalsCount: number;
  achievementsThisWeek: number;
}

export interface WeeklyStats {
  studyTimeMinutes: number;
  studyGoalMinutes: number;
  sessionsCount: number;
  topicsCovered: number;
  masteryImprovement: number;
  flashcardsReviewed: number;
  questionsAnswered: number;
  accuracyPercentage: number | null;
  goalProgressPercentage: number;
}

export interface StrandProgress {
  strand: string;
  strandCode: string | null;
  mastery: number;
  outcomesMastered: number;
  outcomesInProgress: number;
  outcomesTotal: number;
  trend: 'improving' | 'stable' | 'declining';
}

export interface SubjectProgress {
  subjectId: string;
  subjectCode: string;
  subjectName: string;
  subjectColor: string | null;
  masteryLevel: number;
  strandProgress: StrandProgress[];
  recentActivity: string | null;
  sessionsThisWeek: number;
  timeSpentThisWeekMinutes: number;
  xpEarnedThisWeek: number;
  currentFocusOutcomes: string[];
}

export interface FoundationStrength {
  overallStrength: number;
  priorYearMastery: number;
  gapsIdentified: number;
  criticalGaps: string[];
  strengths: string[];
}

export interface StudentProgress {
  studentId: string;
  studentName: string;
  gradeLevel: number;
  schoolStage: string;
  frameworkCode: string | null;
  overallMastery: number;
  foundationStrength: FoundationStrength;
  weeklyStats: WeeklyStats;
  subjectProgress: SubjectProgress[];
  masteryChange30Days: number;
  currentFocusSubjects: string[];
}

export interface InsightItem {
  title: string;
  description: string;
  subjectId: string | null;
  subjectName: string | null;
  priority: 'low' | 'medium' | 'high';
  outcomeCodes: string[];
}

export interface RecommendationItem {
  title: string;
  description: string;
  actionType: string;
  subjectId: string | null;
  estimatedTimeMinutes: number | null;
  priority: 'low' | 'medium' | 'high';
}

export interface PathwayReadiness {
  currentPathway: string;
  recommendedPathway: string | null;
  readyForHigher: boolean;
  blockingGaps: string[];
  strengths: string[];
  recommendation: string;
  confidence: number;
}

export interface HSCProjection {
  predictedBand: number;
  bandRange: string;
  currentAverage: number | null;
  atarContribution: number | null;
  daysUntilHsc: number;
  strengths: string[];
  focusAreas: string[];
  examReadiness: number;
  trajectory: string;
}

export interface WeeklyInsightsData {
  wins: InsightItem[];
  areasToWatch: InsightItem[];
  recommendations: RecommendationItem[];
  teacherTalkingPoints: string[];
  pathwayReadiness: PathwayReadiness | null;
  hscProjection: HSCProjection | null;
  summary: string | null;
}

export interface WeeklyInsights {
  studentId: string;
  weekStart: string;
  insights: WeeklyInsightsData;
  generatedAt: string;
}

export interface Goal {
  id: string;
  studentId: string;
  parentId: string;
  title: string;
  description: string | null;
  targetOutcomes: string[] | null;
  targetMastery: number | null;
  targetDate: string | null;
  reward: string | null;
  isActive: boolean;
  achievedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface GoalProgress {
  currentMastery: number | null;
  progressPercentage: number;
  outcomesMastered: number;
  outcomesTotal: number;
  daysRemaining: number | null;
  isOnTrack: boolean;
}

export interface GoalWithProgress extends Goal {
  progress: GoalProgress;
}

export interface GoalListResponse {
  goals: GoalWithProgress[];
  total: number;
  activeCount: number;
  achievedCount: number;
}

export interface Notification {
  id: string;
  userId: string;
  type: string;
  title: string;
  message: string;
  priority: 'low' | 'normal' | 'high';
  relatedStudentId: string | null;
  relatedSubjectId: string | null;
  relatedGoalId: string | null;
  deliveryMethod: string;
  data: Record<string, unknown> | null;
  createdAt: string;
  sentAt: string | null;
  readAt: string | null;
}

export interface NotificationListResponse {
  notifications: Notification[];
  total: number;
  unreadCount: number;
}

export interface NotificationPreferences {
  userId: string;
  weeklyReports: boolean;
  achievementAlerts: boolean;
  concernAlerts: boolean;
  goalReminders: boolean;
  insightNotifications: boolean;
  emailFrequency: 'daily' | 'weekly' | 'monthly' | 'never';
  preferredTime: string;
  preferredDay: string;
  timezone: string;
  quietHoursStart: string | null;
  quietHoursEnd: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CreateGoalRequest {
  student_id: string;
  title: string;
  description?: string;
  target_outcomes?: string[];
  target_mastery?: number;
  target_date?: string;
  reward?: string;
}

export interface UpdateGoalRequest {
  title?: string;
  description?: string;
  target_outcomes?: string[];
  target_mastery?: number;
  target_date?: string;
  reward?: string;
  is_active?: boolean;
}

export interface UpdateNotificationPreferencesRequest {
  weekly_reports?: boolean;
  achievement_alerts?: boolean;
  concern_alerts?: boolean;
  goal_reminders?: boolean;
  insight_notifications?: boolean;
  email_frequency?: string;
  preferred_time?: string;
  preferred_day?: string;
  timezone?: string;
  quiet_hours_start?: string;
  quiet_hours_end?: string;
}

// =============================================================================
// Response Transformers (with Zod validation)
// =============================================================================

function transformDashboardOverview(response: DashboardOverviewApiResponse): DashboardOverview {
  return {
    students: response.students.map((s) => ({
      id: s.id,
      displayName: s.display_name,
      gradeLevel: s.grade_level,
      schoolStage: s.school_stage,
      frameworkId: s.framework_id,
      totalXp: s.total_xp,
      level: s.level,
      currentStreak: s.current_streak,
      longestStreak: s.longest_streak,
      lastActiveAt: s.last_active_at,
      sessionsThisWeek: s.sessions_this_week,
      studyTimeThisWeekMinutes: s.study_time_this_week_minutes,
    })),
    totalStudyTimeWeekMinutes: response.total_study_time_week_minutes,
    totalSessionsWeek: response.total_sessions_week,
    unreadNotifications: response.unread_notifications,
    activeGoalsCount: response.active_goals_count,
    achievementsThisWeek: response.achievements_this_week,
  };
}

function transformStudentProgress(response: StudentProgressApiResponse): StudentProgress {
  return {
    studentId: response.student_id,
    studentName: response.student_name,
    gradeLevel: response.grade_level,
    schoolStage: response.school_stage,
    frameworkCode: response.framework_code,
    overallMastery: response.overall_mastery,
    foundationStrength: {
      overallStrength: response.foundation_strength.overall_strength,
      priorYearMastery: response.foundation_strength.prior_year_mastery,
      gapsIdentified: response.foundation_strength.gaps_identified,
      criticalGaps: response.foundation_strength.critical_gaps,
      strengths: response.foundation_strength.strengths,
    },
    weeklyStats: {
      studyTimeMinutes: response.weekly_stats.study_time_minutes,
      studyGoalMinutes: response.weekly_stats.study_goal_minutes,
      sessionsCount: response.weekly_stats.sessions_count,
      topicsCovered: response.weekly_stats.topics_covered,
      masteryImprovement: response.weekly_stats.mastery_improvement,
      flashcardsReviewed: response.weekly_stats.flashcards_reviewed,
      questionsAnswered: response.weekly_stats.questions_answered,
      accuracyPercentage: response.weekly_stats.accuracy_percentage,
      goalProgressPercentage: response.weekly_stats.goal_progress_percentage ?? 0,
    },
    subjectProgress: response.subject_progress.map((sp) => ({
      subjectId: sp.subject_id,
      subjectCode: sp.subject_code,
      subjectName: sp.subject_name,
      subjectColor: sp.subject_color,
      masteryLevel: sp.mastery_level,
      strandProgress: sp.strand_progress.map((strand) => ({
        strand: strand.strand,
        strandCode: strand.strand_code,
        mastery: strand.mastery,
        outcomesMastered: strand.outcomes_mastered,
        outcomesInProgress: strand.outcomes_in_progress,
        outcomesTotal: strand.outcomes_total,
        trend: strand.trend,
      })),
      recentActivity: sp.recent_activity,
      sessionsThisWeek: sp.sessions_this_week,
      timeSpentThisWeekMinutes: sp.time_spent_this_week_minutes,
      xpEarnedThisWeek: sp.xp_earned_this_week,
      currentFocusOutcomes: sp.current_focus_outcomes,
    })),
    masteryChange30Days: response.mastery_change_30_days,
    currentFocusSubjects: response.current_focus_subjects,
  };
}

function transformWeeklyInsights(response: WeeklyInsightsApiResponse): WeeklyInsights {
  return {
    studentId: response.student_id,
    weekStart: response.week_start,
    insights: {
      wins: response.insights.wins.map((item) => ({
        title: item.title,
        description: item.description,
        subjectId: item.subject_id,
        subjectName: item.subject_name,
        priority: item.priority,
        outcomeCodes: item.outcome_codes,
      })),
      areasToWatch: response.insights.areas_to_watch.map((item) => ({
        title: item.title,
        description: item.description,
        subjectId: item.subject_id,
        subjectName: item.subject_name,
        priority: item.priority,
        outcomeCodes: item.outcome_codes,
      })),
      recommendations: response.insights.recommendations.map((item) => ({
        title: item.title,
        description: item.description,
        actionType: item.action_type,
        subjectId: item.subject_id,
        estimatedTimeMinutes: item.estimated_time_minutes,
        priority: item.priority,
      })),
      teacherTalkingPoints: response.insights.teacher_talking_points,
      pathwayReadiness: response.insights.pathway_readiness
        ? {
            currentPathway: response.insights.pathway_readiness.current_pathway,
            recommendedPathway: response.insights.pathway_readiness.recommended_pathway,
            readyForHigher: response.insights.pathway_readiness.ready_for_higher,
            blockingGaps: response.insights.pathway_readiness.blocking_gaps,
            strengths: response.insights.pathway_readiness.strengths,
            recommendation: response.insights.pathway_readiness.recommendation,
            confidence: response.insights.pathway_readiness.confidence,
          }
        : null,
      hscProjection: response.insights.hsc_projection
        ? {
            predictedBand: response.insights.hsc_projection.predicted_band,
            bandRange: response.insights.hsc_projection.band_range,
            currentAverage: response.insights.hsc_projection.current_average,
            atarContribution: response.insights.hsc_projection.atar_contribution,
            daysUntilHsc: response.insights.hsc_projection.days_until_hsc,
            strengths: response.insights.hsc_projection.strengths,
            focusAreas: response.insights.hsc_projection.focus_areas,
            examReadiness: response.insights.hsc_projection.exam_readiness,
            trajectory: response.insights.hsc_projection.trajectory,
          }
        : null,
      summary: response.insights.summary,
    },
    generatedAt: response.generated_at,
  };
}

function transformGoalWithProgressValidated(response: GoalWithProgressApiResponse): GoalWithProgress {
  return {
    id: response.id,
    studentId: response.student_id,
    parentId: response.parent_id,
    title: response.title,
    description: response.description,
    targetOutcomes: response.target_outcomes,
    targetMastery: response.target_mastery,
    targetDate: response.target_date,
    reward: response.reward,
    isActive: response.is_active,
    achievedAt: response.achieved_at,
    createdAt: response.created_at,
    updatedAt: response.updated_at,
    progress: {
      currentMastery: response.progress.current_mastery,
      progressPercentage: response.progress.progress_percentage,
      outcomesMastered: response.progress.outcomes_mastered,
      outcomesTotal: response.progress.outcomes_total,
      daysRemaining: response.progress.days_remaining,
      isOnTrack: response.progress.is_on_track,
    },
  };
}

function transformNotificationValidated(response: NotificationApiResponse): Notification {
  return {
    id: response.id,
    userId: response.user_id,
    type: response.type,
    title: response.title,
    message: response.message,
    priority: response.priority,
    relatedStudentId: response.related_student_id,
    relatedSubjectId: response.related_subject_id,
    relatedGoalId: response.related_goal_id,
    deliveryMethod: response.delivery_method,
    data: response.data,
    createdAt: response.created_at,
    sentAt: response.sent_at,
    readAt: response.read_at,
  };
}

function transformNotificationPreferencesValidated(
  response: NotificationPreferencesApiResponse
): NotificationPreferences {
  return {
    userId: response.user_id,
    weeklyReports: response.weekly_reports,
    achievementAlerts: response.achievement_alerts,
    concernAlerts: response.concern_alerts,
    goalReminders: response.goal_reminders,
    insightNotifications: response.insight_notifications,
    emailFrequency: response.email_frequency,
    preferredTime: response.preferred_time,
    preferredDay: response.preferred_day,
    timezone: response.timezone,
    quietHoursStart: response.quiet_hours_start,
    quietHoursEnd: response.quiet_hours_end,
    createdAt: response.created_at,
    updatedAt: response.updated_at,
  };
}

// =============================================================================
// API Client
// =============================================================================

export const parentDashboardApi = {
  /**
   * Get dashboard overview with all children's summaries.
   * Validates response with Zod schema.
   */
  async getDashboard(): Promise<DashboardOverview> {
    const response = await api.get<unknown>('/api/v1/parent/dashboard');
    const validated = DashboardOverviewResponseSchema.parse(response);
    return transformDashboardOverview(validated);
  },

  /**
   * Get detailed progress for a specific student.
   * Validates response with Zod schema.
   */
  async getStudentProgress(studentId: string): Promise<StudentProgress> {
    const response = await api.get<unknown>(
      `/api/v1/parent/students/${studentId}/progress`
    );
    const validated = StudentProgressResponseSchema.parse(response);
    return transformStudentProgress(validated);
  },

  /**
   * Get weekly AI insights for a student.
   * Validates response with Zod schema.
   */
  async getWeeklyInsights(
    studentId: string,
    options?: { weekStart?: string; forceRegenerate?: boolean }
  ): Promise<WeeklyInsights> {
    const params = new URLSearchParams();
    if (options?.weekStart) params.set('week_start', options.weekStart);
    if (options?.forceRegenerate) params.set('force_regenerate', 'true');

    const url = `/api/v1/parent/students/${studentId}/insights${params.toString() ? `?${params}` : ''}`;
    const response = await api.get<unknown>(url);
    const validated = WeeklyInsightsResponseSchema.parse(response);
    return transformWeeklyInsights(validated);
  },

  /**
   * List goals (optionally filtered by student).
   * Validates response with Zod schema.
   */
  async getGoals(options?: {
    studentId?: string;
    activeOnly?: boolean;
    page?: number;
    pageSize?: number;
  }): Promise<GoalListResponse> {
    const params = new URLSearchParams();
    if (options?.studentId) params.set('student_id', options.studentId);
    if (options?.activeOnly) params.set('active_only', 'true');
    if (options?.page) params.set('page', options.page.toString());
    if (options?.pageSize) params.set('page_size', options.pageSize.toString());

    const url = `/api/v1/parent/goals${params.toString() ? `?${params}` : ''}`;
    const response = await api.get<unknown>(url);
    const validated = GoalListResponseSchema.parse(response);

    return {
      goals: validated.goals.map(transformGoalWithProgressValidated),
      total: validated.total,
      activeCount: validated.active_count ?? 0,
      achievedCount: validated.achieved_count ?? 0,
    };
  },

  /**
   * Create a new goal.
   * Validates response with Zod schema.
   */
  async createGoal(data: CreateGoalRequest): Promise<GoalWithProgress> {
    const response = await api.post<unknown>('/api/v1/parent/goals', data);
    const validated = GoalWithProgressResponseSchema.parse(response);
    return transformGoalWithProgressValidated(validated);
  },

  /**
   * Update a goal.
   * Validates response with Zod schema.
   */
  async updateGoal(goalId: string, data: UpdateGoalRequest): Promise<GoalWithProgress> {
    const response = await api.put<unknown>(`/api/v1/parent/goals/${goalId}`, data);
    const validated = GoalWithProgressResponseSchema.parse(response);
    return transformGoalWithProgressValidated(validated);
  },

  /**
   * Delete a goal.
   */
  async deleteGoal(goalId: string): Promise<void> {
    await api.delete(`/api/v1/parent/goals/${goalId}`);
  },

  /**
   * Check if a goal has been achieved.
   * Validates response with Zod schema.
   */
  async checkGoalAchievement(goalId: string): Promise<GoalWithProgress> {
    const response = await api.post<unknown>(
      `/api/v1/parent/goals/${goalId}/check-achievement`
    );
    const validated = GoalWithProgressResponseSchema.parse(response);
    return transformGoalWithProgressValidated(validated);
  },

  /**
   * Get notifications.
   * Validates response with Zod schema.
   */
  async getNotifications(options?: {
    unreadOnly?: boolean;
    type?: string;
    page?: number;
    pageSize?: number;
  }): Promise<NotificationListResponse> {
    const params = new URLSearchParams();
    if (options?.unreadOnly) params.set('unread_only', 'true');
    if (options?.type) params.set('notification_type', options.type);
    if (options?.page) params.set('page', options.page.toString());
    if (options?.pageSize) params.set('page_size', options.pageSize.toString());

    const url = `/api/v1/parent/notifications${params.toString() ? `?${params}` : ''}`;
    const response = await api.get<unknown>(url);
    const validated = NotificationListResponseSchema.parse(response);

    return {
      notifications: validated.notifications.map(transformNotificationValidated),
      total: validated.total,
      unreadCount: validated.unread_count,
    };
  },

  /**
   * Mark a notification as read.
   * Validates response with Zod schema.
   */
  async markNotificationRead(notificationId: string): Promise<Notification> {
    const response = await api.post<unknown>(
      `/api/v1/parent/notifications/${notificationId}/read`
    );
    const validated = NotificationResponseSchema.parse(response);
    return transformNotificationValidated(validated);
  },

  /**
   * Mark all notifications as read.
   */
  async markAllNotificationsRead(): Promise<{ markedRead: number }> {
    const response = await api.post<{ marked_read: number }>(
      '/api/v1/parent/notifications/read-all'
    );
    return { markedRead: response.marked_read };
  },

  /**
   * Get notification preferences.
   * Validates response with Zod schema.
   */
  async getNotificationPreferences(): Promise<NotificationPreferences> {
    const response = await api.get<unknown>(
      '/api/v1/parent/notification-preferences'
    );
    const validated = NotificationPreferencesResponseSchema.parse(response);
    return transformNotificationPreferencesValidated(validated);
  },

  /**
   * Update notification preferences.
   * Validates response with Zod schema.
   */
  async updateNotificationPreferences(
    data: UpdateNotificationPreferencesRequest
  ): Promise<NotificationPreferences> {
    const response = await api.put<unknown>(
      '/api/v1/parent/notification-preferences',
      data
    );
    const validated = NotificationPreferencesResponseSchema.parse(response);
    return transformNotificationPreferencesValidated(validated);
  },
};

File: backend/app/schemas/parent_dashboard.py
"""Parent dashboard schemas for progress overview and analytics."""
from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import Field

from app.schemas.base import BaseSchema
from app.schemas.weekly_insight import WeeklyInsightsData


# =============================================================================
# Student Summary for Dashboard
# =============================================================================


class DashboardStudentSummary(BaseSchema):
    """Student summary for parent dashboard."""

    id: UUID
    display_name: str
    grade_level: int
    school_stage: str
    framework_id: UUID | None = None
    # Gamification stats
    total_xp: int = 0
    level: int = 1
    current_streak: int = 0
    longest_streak: int = 0
    # Activity
    last_active_at: datetime | None = None
    sessions_this_week: int = 0
    study_time_this_week_minutes: int = 0


# =============================================================================
# Progress Metrics
# =============================================================================


class WeeklyStats(BaseSchema):
    """Weekly study statistics."""

    study_time_minutes: int = Field(0, ge=0, description="Total study time in minutes")
    study_goal_minutes: int = Field(
        150, ge=0, description="Weekly study goal in minutes (default 2.5 hrs)"
    )
    sessions_count: int = Field(0, ge=0, description="Number of study sessions")
    topics_covered: int = Field(0, ge=0, description="Unique topics/outcomes worked on")
    mastery_improvement: Decimal = Field(
        default=Decimal("0"), description="Mastery change this week (can be negative)"
    )
    flashcards_reviewed: int = Field(0, ge=0, description="Flashcards reviewed")
    questions_answered: int = Field(0, ge=0, description="Questions answered in sessions")
    accuracy_percentage: Decimal | None = Field(
        None, ge=0, le=100, description="Answer accuracy percentage"
    )

    @property
    def goal_progress_percentage(self) -> Decimal:
        """Calculate progress towards weekly goal."""
        if self.study_goal_minutes <= 0:
            return Decimal("100")
        return min(
            Decimal("100"),
            Decimal(str(self.study_time_minutes * 100 / self.study_goal_minutes)),
        )


class StrandProgress(BaseSchema):
    """Progress within a curriculum strand."""

    strand: str = Field(..., description="Strand name (e.g., 'Number & Algebra')")
    strand_code: str | None = Field(None, description="Strand code if available")
    mastery: Decimal = Field(..., ge=0, le=100, description="Mastery percentage")
    outcomes_mastered: int = Field(0, ge=0, description="Outcomes at mastery level")
    outcomes_in_progress: int = Field(0, ge=0, description="Outcomes being worked on")
    outcomes_total: int = Field(0, ge=0, description="Total outcomes in strand")
    trend: Literal["improving", "stable", "declining"] = Field(
        default="stable", description="Recent trend"
    )


class SubjectProgress(BaseSchema):
    """Progress within a subject."""

    subject_id: UUID
    subject_code: str
    subject_name: str
    subject_color: str | None = None
    mastery_level: Decimal = Field(..., ge=0, le=100, description="Overall mastery")
    strand_progress: list[StrandProgress] = Field(
        default_factory=list, description="Progress by strand"
    )
    recent_activity: datetime | None = Field(None, description="Last activity in subject")
    sessions_this_week: int = 0
    time_spent_this_week_minutes: int = 0
    xp_earned_this_week: int = 0
    current_focus_outcomes: list[str] = Field(
        default_factory=list, description="Outcomes currently being focused on"
    )


class FoundationStrength(BaseSchema):
    """Foundation strength assessment."""

    overall_strength: Decimal = Field(
        ..., ge=0, le=100, description="Overall foundation strength percentage"
    )
    prior_year_mastery: Decimal = Field(
        ..., ge=0, le=100, description="Mastery of prior year outcomes"
    )
    gaps_identified: int = Field(0, ge=0, description="Number of foundation gaps")
    critical_gaps: list[str] = Field(
        default_factory=list, description="Critical gap descriptions"
    )
    strengths: list[str] = Field(
        default_factory=list, description="Foundation strengths"
    )


# =============================================================================
# Dashboard Overview Response
# =============================================================================


class DashboardOverviewResponse(BaseSchema):
    """Main parent dashboard overview."""

    # Students
    students: list[DashboardStudentSummary] = Field(
        ..., description="Summary of all children"
    )
    # Aggregate stats
    total_study_time_week_minutes: int = Field(
        0, description="Combined study time this week"
    )
    total_sessions_week: int = Field(0, description="Combined sessions this week")
    # Notifications
    unread_notifications: int = Field(0, ge=0, description="Unread notification count")
    # Quick insights
    active_goals_count: int = Field(0, ge=0, description="Number of active goals")
    achievements_this_week: int = Field(0, ge=0, description="Achievements unlocked this week")


# =============================================================================
# Student Progress Response
# =============================================================================


class StudentProgressResponse(BaseSchema):
    """Detailed progress for a single student."""

    student_id: UUID
    student_name: str
    grade_level: int
    school_stage: str
    framework_code: str | None = None
    # Overall metrics
    overall_mastery: Decimal = Field(..., ge=0, le=100, description="Overall curriculum mastery")
    foundation_strength: FoundationStrength
    # Weekly breakdown
    weekly_stats: WeeklyStats
    # Subject breakdown
    subject_progress: list[SubjectProgress] = Field(
        default_factory=list, description="Progress by subject"
    )
    # Comparison
    mastery_change_30_days: Decimal = Field(
        default=Decimal("0"), description="Mastery change over last 30 days"
    )
    # Current focus
    current_focus_subjects: list[str] = Field(
        default_factory=list, description="Subjects currently being focused on"
    )


# =============================================================================
# Subject Detail Response
# =============================================================================


class SubjectProgressDetailResponse(BaseSchema):
    """Detailed progress for a single subject."""

    student_id: UUID
    student_name: str
    subject_id: UUID
    subject_code: str
    subject_name: str
    subject_color: str | None = None
    # Mastery
    overall_mastery: Decimal = Field(..., ge=0, le=100)
    strand_progress: list[StrandProgress]
    # Activity
    weekly_stats: WeeklyStats
    recent_sessions: list[dict[str, Any]] = Field(
        default_factory=list, description="Recent session summaries"
    )
    # Outcomes
    outcomes_mastered: int = 0
    outcomes_in_progress: int = 0
    outcomes_not_started: int = 0
    outcomes_total: int = 0
    # Focus outcomes
    current_focus_outcomes: list[dict[str, Any]] = Field(
        default_factory=list, description="Current focus outcomes with details"
    )
    # Pathway (Stage 5 Maths only)
    pathway: str | None = Field(None, description="5.1, 5.2, or 5.3 for Stage 5 Maths")
    # HSC (Stage 6 only)
    hsc_course_name: str | None = None
    predicted_band: int | None = Field(None, ge=1, le=6)


# =============================================================================
# Weekly Summary (for email)
# =============================================================================


class WeeklySummaryResponse(BaseSchema):
    """Weekly summary for email delivery."""

    student_id: UUID
    student_name: str
    week_start: date
    week_end: date
    # Stats
    weekly_stats: WeeklyStats
    # Subject highlights
    top_subjects: list[dict[str, Any]] = Field(
        default_factory=list, description="Top performing subjects"
    )
    needs_attention: list[dict[str, Any]] = Field(
        default_factory=list, description="Subjects needing attention"
    )
    # Insights
    insights: WeeklyInsightsData | None = None
    # Goals
    goals_progress: list[dict[str, Any]] = Field(
        default_factory=list, description="Active goals progress"
    )
    goals_achieved_this_week: list[dict[str, Any]] = Field(
        default_factory=list, description="Goals achieved this week"
    )
    # Achievements
    achievements_unlocked: list[str] = Field(
        default_factory=list, description="Achievements unlocked this week"
    )
    # Actions
    suggested_actions: list[str] = Field(
        default_factory=list, description="Suggested actions for parents"
    )


class SendWeeklySummaryRequest(BaseSchema):
    """Request to send weekly summary email."""

    student_id: UUID
    week_start: date | None = Field(
        None, description="Week to send summary for (defaults to last week)"
    )
    send_now: bool = Field(
        True, description="Send immediately or queue for preferred time"
    )


class SendWeeklySummaryResponse(BaseSchema):
    """Response after sending weekly summary."""

    success: bool
    message: str
    email_sent_to: str | None = None
    sent_at: datetime | None = None
